feat: new linting rules for compositions keys oneOf, anyOf, allOf (#2369)
Redocly/redocly-cli: docs/@v2/rules/oas/no-illogical-composition-keywords.md

---
slug: /docs/cli/rules/oas/no-illogical-composition-keywords
---

# no-illogical-composition-keywords

Ensures that `oneOf`, `anyOf`, and `allOf` combine schemas that a value can actually resolve against.

The rule reports:

- A `oneOf` or `anyOf` with fewer than two schemas, unless the schema declares a `discriminator`.
- An `allOf` with fewer than two schemas that neither declares another keyword of its own nor extends a discriminated schema.
- A schema repeated inside the same keyword.
- An empty schema (`{}`) used as a member.
- Two `oneOf` schemas that a single value can match at the same time.
- A nullable schema whose `oneOf` also accepts `null`.
- A `discriminator` whose property is missing from `required` in every member schema.

> - An inline `oneOf` or `anyOf` member that a `discriminator` cannot select.

Reviewer: ```suggestion
- `oneOf` or `anyOf` with fewer than two schemas, unless the schema declares a `discriminator`
- `allOf` with fewer than two schemas that neither declares another keyword of its own nor extends a discriminated schema
- schemas repeated inside the same keyword
- empty schemas (`{}`) used as members
- two `oneOf` schemas that a single value can match at the same time
- nullable schemas whose `oneOf` also accepts `null`
- discriminators whose property is missing from `required` in every member schema
- inline `oneOf` or `anyOf` members that a `discriminator` cannot select
```

Author: Two points looks wrong for me:
1. "nullable schemas whose `oneOf` also accepts `null`." - I mean, that if parent schema has a `null` and the sibling schema inside `oneOf` also accepts null we report a warning. So, i will say the proper sentence sounds:
> - schemas that accept null alongside a oneOf member that also accepts null

2. "discriminators whose property is missing from `required` in every member schema" - `discriminators` word looks weird for me, because we have one discriminator in one schema, that accepts one prop. So, the proper sentence sounds:
> - member schemas that omit the discriminator property from required

Reviewer: Thank you! Updated and applied.

> | OAS | Compatibility |

Reviewer 2: I guess the same could be extended later to the AsyncAPI spec.

Author: Yeap, we can do it, but need more investigation with the specific cases usage of those keywords in Async.

Reviewer 2: Sure, not in this PR.

```

## API design principles

`oneOf` means "exactly one".
When a value matches two of the listed schemas, no tool can tell which one was intended, and validators, code generators, and documentation all disagree about the result.

The most common version of this is nullability: if a referenced schema already accepts `null` and the `oneOf` also lists `type: 'null'`, a null value matches both branches.
The same ambiguity appears one level up, when the schema holding the `oneOf` is itself nullable and a member accepts `null` too.

Deciding whether two arbitrary schemas overlap is not solvable in general, so the comparison stays deliberately narrow.
It reads `type`, `nullable`, `enum`, `const`, `properties`, `required`, and `additionalProperties: false`.
A `const` counts as a single-value `enum`, so one member can use `enum` and the other `const`.
When a member uses any other constraint, such as `not`, `pattern`, `minimum`, or a nested `allOf`, that constraint may be what separates the schemas, so the rule reports nothing for the pair.

A `discriminator` names the property that tells the members apart, so the rule trusts it and checks only what the specification requires.

> The property must be listed in `required` in every member schema, because a value can otherwise omit it and nothing decides which schema applies.

Reviewer 2: I believe this is already covered by another rule (https://redocly.com/docs/cli/rules/oas/spec-discriminator-defaultMapping), at least for OAS 3.2 (where it's a specification requirement).

Author: Good catch, let me investigate this.

Author: Thanks for the suggestion. I added support of `defaultMapping` to the current rule. I am also considered about duplication between spec rule and this rule and decided to keep both, because we don't know if user use both rules or just `no-illogical-composition-keywords` alone. If user will have two duplicated errors he can just fix one issue and remove two errors or add these to lint ignore.

The rule reports:

- A `oneOf` or `anyOf` with fewer than two schemas, unless the schema declares a `discriminator`.
- An `allOf` with fewer than two schemas that neither declares another keyword of its own nor extends a discriminated schema.
- A schema repeated inside the same keyword.
- An empty schema (`{}`) used as a member.
- Two `oneOf` schemas that a single value can match at the same time.
- A nullable schema whose `oneOf` also accepts `null`.
- A `discriminator` whose property is missing from `required` in every member schema.
- An inline `oneOf` or `anyOf` member that a `discriminator` cannot select.

| OAS | Compatibility |
| --- | ------------- |
| 2.0 | ❌            |
| 3.0 | ✅            |
| 3.1 | ✅            |
| 3.2 | ✅            |

```mermaid
flowchart TD

Root ==> components --> NamedSchemas --> Schema

Schema ==> OneOf
Schema ==> AnyOf
Schema ==> AllOf

style OneOf fill:#codaf9,stroke:#0044d4,stroke-width:5px
style AnyOf fill:#codaf9,stroke:#0044d4,stroke-width:5px
style AllOf fill:#codaf9,stroke:#0044d4,stroke-width:5px
```

## API design principles

`oneOf` means "exactly one".

> When a value matches two of the listed schemas, no tool can tell which one was intended, and validators, code generators, and documentation all disagree about the result.

Reviewer: ```suggestion
When a value matches two of the listed schemas, no tool can tell which one was intended.
Validators, code generators, and documentation all disagree about the result.
```

> The most common version of this is nullability: if a referenced schema already accepts `null` and the `oneOf` also lists `type: 'null'`, a null value matches both branches.

Reviewer: ```suggestion
The most common version of this issue is nullability: if a referenced schema already accepts `null` and the `oneOf` also lists `type: 'null'`, a null value matches both branches.
```

An example configuration:

```yaml
rules:
  no-illogical-composition-keywords: error
```

## Examples

Given this configuration:

```yaml
rules:
  no-illogical-composition-keywords: error
```

Example of an **incorrect** `oneOf` where both schemas accept `null`:

```yaml
components:
  schemas:
    TimeShift:
      type: [object, 'null']
    Invoice:
      oneOf:
        - $ref: '#/components/schemas/TimeShift'
        - type: 'null'
```

Example of a **correct** `oneOf`:

```yaml
components:
  schemas:
    TimeShift:
      type: object
    Invoice:
      oneOf:
        - $ref: '#/components/schemas/TimeShift'
        - type: 'null'
```

Example of an **incorrect** `discriminator` with an inline member:

```yaml
components:
  schemas:
    Cat:
      type: object
      properties:
        petType:
          type: string
      required: [petType]
    Pet:
      discriminator:
        propertyName: petType
      oneOf:
        - $ref: '#/components/schemas/Cat'
        - type: object
          properties:
            petType:
              type: string
          required: [petType]
```

> > Move the inline schema into `components/schemas` and reference it with a `$ref`.

Reviewer: ```suggestion
Move the inline schema into `components/schemas` and reference it with a `$ref`.
```

Example of an **incorrect** `discriminator`, where `petType` is optional:

```yaml
components:
  schemas:
    Pet:
      discriminator:
        propertyName: petType
      oneOf:
        - $ref: '#/components/schemas/Cat'
        - $ref: '#/components/schemas/Dog'
    Cat:
      type: object
      properties:
        petType:
          type: string
    Dog:
      type: object
      properties:
        petType:
          type: string
```

> > Add `petType` to `required` in both `Cat` and `Dog` to fix this.

Reviewer: ```suggestion
Add `petType` to `required` in both `Cat` and `Dog` to fix this.
```

Example of a **correct** single-schema `allOf` that declares a subtype:

```yaml
components:
  schemas:
    Pet:
      type: object
      required: [petType]
      properties:
        petType:
          type: string
      discriminator:
        propertyName: petType
    Cat:
      allOf:
        - $ref: '#/components/schemas/Pet'
```

Example of **incorrect** composition keywords:

```yaml
components:
  schemas:
    Pet:
      oneOf:
        - $ref: '#/components/schemas/Cat'
    Animal:
      allOf:
        - $ref: '#/components/schemas/Cat'
        - $ref: '#/components/schemas/Cat'
        - {}
```

> > `Pet` wraps a single schema, and `Animal` repeats one schema and adds an empty one that matches any value.

Reviewer: ```suggestion
`Pet` wraps a single schema, and `Animal` repeats one schema and adds an empty one that matches any value.
```

- A schema repeated inside the same keyword.
- An empty schema (`{}`) used as a member.
- Two `oneOf` schemas that a single value can match at the same time.
- A nullable schema whose `oneOf` also accepts `null`.
- A `discriminator` whose property is missing from `required` in every member schema, on OAS 3.0 and 3.1.
  To cover OAS 3.2 as well, use this rule together with [spec-discriminator-defaultMapping](./spec-discriminator-defaultMapping.md).
- An inline `oneOf` or `anyOf` member that a `discriminator` cannot select.

| OAS | Compatibility |
| --- | ------------- |
| 2.0 | ❌            |
| 3.0 | ✅            |
| 3.1 | ✅            |
| 3.2 | ✅            |

```mermaid
flowchart TD

Root ==> components --> NamedSchemas --> Schema

Schema ==> OneOf
Schema ==> AnyOf
Schema ==> AllOf

style OneOf fill:#codaf9,stroke:#0044d4,stroke-width:5px
style AnyOf fill:#codaf9,stroke:#0044d4,stroke-width:5px
style AllOf fill:#codaf9,stroke:#0044d4,stroke-width:5px
```

## API design principles

`oneOf` means "exactly one".
When a value matches two of the listed schemas, no tool can tell which one was intended, and validators, code generators, and documentation all disagree about the result.

> The most common version of this is nullability: if a referenced schema already accepts `null` and the `oneOf` also lists `type: 'null'`, a null value matches both branches.

Reviewer: ```suggestion
One of the most common ambiguous patterns is nullability.
If a referenced schema already accepts `null` and the `oneOf` also lists `type: 'null'`, a null value matches both branches.
```

The same ambiguity appears one level up, when the schema holding the `oneOf` is itself nullable and a member accepts `null` too.

Deciding whether two arbitrary schemas overlap is not solvable in general, so the comparison stays deliberately narrow.
It reads `type`, `nullable`, `enum`, `const`, `properties`, `required`, and `additionalProperties: false`.

> A `const` counts as a single-value `enum`, so one member can use `enum` and the other `const`.

Reviewer: ```suggestion
A `const` counts as a single-value `enum`.
One member can use `enum` and the other `const`.
```

> When a member uses any other constraint, such as `not`, `pattern`, `minimum`, or a nested `allOf`, that constraint may be what separates the schemas, so the rule reports nothing for the pair.

Reviewer: ```suggestion
When a member uses any other constraint, such as `not`, `pattern`, `minimum`, or a nested `allOf`, that constraint may be what separates the schemas.
In such cases, the rule reports nothing for the pair.
```

> A `discriminator` names the property that tells the members apart, so the rule trusts it and checks only what the specification requires.

Reviewer: ```suggestion
A `discriminator` names the property that tells the members apart.
The rule trusts it and checks only what the specification requires.
```

## Examples

Given this configuration:

```yaml
rules:
  no-illogical-composition-keywords: error
```

Example of an **incorrect** `oneOf` where both schemas accept `null`:

```yaml
components:
  schemas:
    TimeShift:
      type: [object, 'null']
    Invoice:
      oneOf:
        - $ref: '#/components/schemas/TimeShift'
        - type: 'null'
```

Example of a **correct** `oneOf`:

```yaml
components:
  schemas:
    TimeShift:
      type: object
    Invoice:
      oneOf:
        - $ref: '#/components/schemas/TimeShift'
        - type: 'null'
```

Example of an **incorrect** `discriminator` with an inline member:

```yaml
components:
  schemas:
    Cat:
      type: object
      properties:
        petType:
          type: string
      required: [petType]
    Pet:
      discriminator:
        propertyName: petType
      oneOf:
        - $ref: '#/components/schemas/Cat'
        - type: object
          properties:
            petType:
              type: string
          required: [petType]
```

> Move the inline schema into `components/schemas` and reference it with a `$ref`.

Example of an **incorrect** `discriminator`, where `petType` is optional:

```yaml
components:
  schemas:

>     Pet:

Reviewer 2: Could you use Cafe example instead?

Author: Replaced in docs and unit tests.

        petType:
          type: string
```

> Add `petType` to `required` in both `Cat` and `Dog` to fix this.

Example of a **correct** single-schema `allOf` that declares a subtype:

```yaml
components:
  schemas:
    Pet:
      type: object
      required: [petType]
      properties:
        petType:
          type: string
      discriminator:
        propertyName: petType
    Cat:
      allOf:
        - $ref: '#/components/schemas/Pet'
```

Example of **incorrect** composition keywords:

```yaml
components:
  schemas:
    Pet:
      oneOf:
        - $ref: '#/components/schemas/Cat'
    Animal:
      allOf:
        - $ref: '#/components/schemas/Cat'
        - $ref: '#/components/schemas/Cat'
        - {}
```

> `Pet` wraps a single schema, and `Animal` repeats one schema and adds an empty one that matches any value.

## Related rules

- [no-schema-type-mismatch](../common/no-schema-type-mismatch.md)
- [no-required-schema-properties-undefined](../common/no-required-schema-properties-undefined.md)
- [spec-discriminator-defaultMapping](./spec-discriminator-defaultMapping.md) —
  reports an optional discriminator property on OAS 3.2, where the specification requires it.

>   This rule covers OAS 3.0 and 3.1 instead, so enable both to check every version.

Reviewer: ```suggestion
- [spec-discriminator-defaultMapping](./spec-discriminator-defaultMapping.md) — on OAS 3.2, flags a discriminator whose `propertyName` is optional but has no `defaultMapping` required by OAS 3.2.
  `no-illogical-composition-keywords` checks the same issue on OAS 3.0 and 3.1, where `defaultMapping` doesn't exist.
  Enable both rules to cover every OAS version.
```

## Resources

> 

Reviewer 2: Maybe mention the corresponding Adam's article here?

Author: Thanks, added.

| -------- | ------ | ----------------------------------------------------------------------------------------- |
| severity | string | Possible values: `off`, `warn`, `error`. Default `warn` (in `recommended` configuration). |

An example configuration:

```yaml
rules:
  no-illogical-composition-keywords: error
```

## Examples

Given this configuration:

```yaml
rules:
  no-illogical-composition-keywords: error
```

Example of an **incorrect** `oneOf` where both schemas accept `null`:

```yaml
components:
  schemas:
    PhotoUrl:
      type: [string, 'null']
      format: uri
    MenuItemPhoto:
      oneOf:
        - $ref: '#/components/schemas/PhotoUrl'
        - type: 'null'
```

Example of a **correct** `oneOf`:

```yaml
components:
  schemas:
    PhotoUrl:
      type: string
      format: uri
    MenuItemPhoto:
      oneOf:
        - $ref: '#/components/schemas/PhotoUrl'
        - type: 'null'
```

Example of an **incorrect** `discriminator` with an inline member:

```yaml
components:
  schemas:
    Beverage:
      type: object
      properties:
        category:
          type: string
          const: beverage

>       required: [category]

Reviewer 2: I'd suggest against mixing json/yaml syntax in the examples unless necessary. Albeit they are valid, it's better to stick to the most consistent approach in examples to confuse users less.
